contributor-docs/adrs/adr-021-automated-prop-docs.md

```diff
@@ -0,0 +1,277 @@
+- Customizable through [plugins](https://typedoc.org/guides/plugins/)
+
+##### TypeDoc cons
+
+- I haven't been able to parse the TypeDoc output into a `components.json` file (example of existing JSON file: [components.json](https://unpkg.com/browse/@primer/react@36.27.0/generated/components.json))
+- I'm not able to get it to work without some updates to our `tsconfig` that I'm not sure we actually want
+- Props on components cast using `as PolymorphicForwardRefComponent<El, Props>` are not picked up
+- Not well-documented and hard to debug
+- There are no existing plugins to help document React components
+- Seems like the point of this tool is to generate a static site for documentation. It's less useful for extracting type data and doing something custom with it.
+- I'm not sure how to print the returned statements/expressions (in the format of a TypeScript Symbol) in a way that will make it easier to debug. `JSON.stringify()` fails due to circular references.
+
+## Impact
+
+- We don't need to manually create `{ComponentName}.docs.json` for each component files for each component
+- Component metadata such as prop types and their descriptions will always be in sync with what's in the source code
+  - We only use this data for component docs on primer.style right now ([DataTable example](https://primer.style/components/data-table/react/draft#props)), but the possibilities are endless. This may even be useful for keeping Figma components in sync.
+- More thorough JSDoc comments provide more context when:
+  - working directly in PRC
+  - using PRC imports in an IDE
+
+## Decision
+
+TBD:
+- Will we move forward with automating prop docs?
+- If so, how? What are the suggested tools and process?
+
+### Potential paths forward
+
+- Move forward with `react-typescript-docgen` (preferred)
+  - Could be handled by @mperrotti with minimal help from an engineer
+- Move forward with TypeDoc
+  - Would require more time from one or more engineers
+- Build our own tooling to extract API data from component source code
+  - Would take a significant amount of high-effort work and would require more assistance from one or more engineer who are well-versed in TypeScript and ASTs
+- Manually maintain `.docs.json` files temporarily, figure out automation later
+  - If we're not comfortable with the previous directions, we could punt on this ADR and spend time updating `.docs.json` files for accuracy
+- Continue to manually maintain `.docs.json` files long-term
+  - Keep handling component API docs as we do today, but update this ADR to document how and why we document components this way
+
+
+## Open questions
+
+- Do we want to generate docs for internal component props? If so, where would they be rendered?
+- What should we do about hooks? It would be useful to have their data for docs, but it might be strange to keep that data in a file named **components**.json
```

Comment:
Agreed, there definitely is a bigger concept at play here that is basically "api.json" that captures components, functions, constants, etc from a JavaScript package.